File: VotingSystem/test/voting/VoterTest.java
package voting;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class VoterTest {

    @Test
    public void testThatCandidateIdIsCorrect() {
        ElectoralBody electoralBody = new ElectoralBody();
        Voter voter1 = electoralBody.registerVoter("Ade", "Wale", "correctPassword");
        assertEquals(electoralBody.getVoterId(), voter1.getVoterId());
        Voter voter2 = electoralBody.registerVoter("Ade", "Wale", "correctPassword");
        assertEquals(electoralBody.getVoterId(), voter2.getVoterId());
        Voter voter3 = electoralBody.registerVoter("Ade", "Wale", "correctPassword");
        assertEquals(electoralBody.getVoterId(), voter3.getVoterId());
    }

    @Test
    public void testThatUpdatePasswordMethodThrowsExceptionForIncorrectPassword() {
        Voter voter1 = new Voter("Ali", "correctPassword", 0000000000);
        String newPassword = "newPassword";
        IllegalArgumentException illegalArgumentException = assertThrows(IllegalArgumentException.class, () -> {voter1.updatePassword("IncorrectPassword", newPassword);});
        assertEquals(illegalArgumentException.getMessage(), "Invalid credentials");
    }

    @Test
    public void testThatExceptionIsThrownIfVoterTriesToVoterForACandidateMoreThanOnce() {
        ElectoralBody electoralBody = new ElectoralBody();
        Candidate candidate1 = electoralBody.registerCandidate("ade", "wale", "President");
        Candidate candidate2 = electoralBody.registerCandidate("ade", "wale", "President");
        Candidate candidate3 = electoralBody.registerCandidate("ade", "wale", "President");
        Voter voter = electoralBody.registerVoter("ade", "wale", "1234");
        voter.castVote(electoralBody.getVoterId(), 1);
        CandidateHasBeenVotedException candidateHasBeenVotedException = assertThrows(CandidateHasBeenVotedException.class, () -> {voter.castVote(electoralBody.getVoterId(), 1);});
        assertEquals(candidateHasBeenVotedException.getMessage(), "You have voted already voted for this candidate");

    }

    @Test
    public void testThatExceptionIsThrownIfVoterIdDoesNotMatch() {
        ElectoralBody electoralBody = new ElectoralBody();
        Candidate candidate1 = electoralBody.registerCandidate("ade", "wale", "President");
        Voter voter = electoralBody.registerVoter("ade", "wale", "1234");

        VoterNotRegisteredException voterNotRegisteredException = assertThrows(VoterNotRegisteredException.class, () -> {voter.castVote(2, 1);});
        assertEquals(voterNotRegisteredException.getMessage(), "Voter with id 2 is not registered");

    }
    @Test
    public void testThatExceptionIsThrownIfCandidateIdDoesNotMatch() {
        ElectoralBody electoralBody = new ElectoralBody();
        Candidate candidate1 = electoralBody.registerCandidate("ade", "wale", "President");
        Voter voter = electoralBody.registerVoter("ade", "wale", "1234");

        CandidateNotRegisteredException candidateNotRegisteredException = assertThrows(CandidateNotRegisteredException.class, () -> {voter.validateCandidate(-1);});
        assertEquals(candidateNotRegisteredException.getMessage(), "Candidate with id -1 is not registered");
    }



}

File: VotingServiceJs/Voter.test.js
const ElectoralBody = require("./ElectoralBody");


test("test that exception is thrown if voter has voted for a specific candidate", () => {
    let electoralBody = new ElectoralBody();
    let candidate1 = electoralBody.registerCandidate("willy", "willy", "President");
    let voter1 = electoralBody.registerVoter("oga", "oga", "1234");
    voter1.castVote(candidate1.id,voter1.id);
    let vote =()=> voter1.castVote(candidate1.id,voter1.id);
    let errorMessage = "Can't vote twice";
    expect(vote).toThrowError(errorMessage);
})

test("test that exception is thrown if candidate is not registered", () => {
    let electoralBody = new ElectoralBody();
    let candidate1 =  electoralBody.registerCandidate("willy", "willy", "President");
    let voter1 = electoralBody.registerVoter("oga", "oga", "1234");
    let vote =()=>voter1.castVote(2,voter1.id);
    let errorMessage = "Candidate not registered";
    expect(vote).toThrowError(errorMessage);

})

test("test that exception is thrown if voter is not registered",()=>{
    let electoralBody = new ElectoralBody();
    let candidate1 = electoralBody.registerCandidate("willy", "willy", "President");
    let voter1 = electoralBody.registerVoter("oga", "oga", "1234");
    let vote =()=>voter1.castVote(candidate1.id,1);
    let errorMessage = "Voter not registered";
    expect(vote).toThrowError(errorMessage);
})
